Week-4/exercise4_1.js

```diff
@@ -0,0 +1,119 @@
+const STATE = {
+    FULFILLED: "fulfilled",
+    REJECTED: "rejected",
+    PENDING: "pending"
+  };
+  class MyPromise {
+    constructor(cb) {
+      this.value;
+      this.state = STATE.PENDING;
+      this.thenCbs = [];
+      this.catchCbs = [];
+  
+      this.runCallBacks = () => {
+        if (this.state === STATE.FULFILLED) {
+          this.thenCbs.forEach((callBack) => {
+            callBack(this.value);
+          });
+          this.thenCbs = [];
+        }
+  
+        if (this.state === STATE.REJECTED) {
+          this.catchCbs.forEach((callBack) => {
+            callBack(this.value);
+          });
+          this.catchCbs = [];
+        }
+      };
+  
+      const onSuccess = (value) => {
+        if (this.state !== STATE.PENDING) {
+          return;
+        } else {
+          this.value = value;
+          this.state = STATE.FULFILLED;
+          this.runCallBacks();
+        }
+      };
+      const onFail = (value) => {
+        if (this.state !== STATE.PENDING) {
+          return;
+        } else {
+          this.value = value;
+          this.state = STATE.REJECTED;
+          this.runCallBacks();
+        }
+      };
+  
+      try {
+        cb(onSuccess, onFail);
+      } catch (err) {
+        onFail(err);
+      }
+    }
+  
+    then(thenCb, catchCb) {
```

Comment:
Try adding comments everywhere. Build this as a habit when you are writing a code for abstract layer like this